src/Assembly/Helpers/Net/Sockets/NetworkPokeServer.cs

```
@@ -0,0 +1,211 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net;
+using System.Net.Sockets;
+using System.IO;
+using System.Collections.ObjectModel;
+using Mono.Nat;
+
+namespace Assembly.Helpers.Net.Sockets
+{
+	/// <summary>
+	/// Network poking server.
+	/// </summary>
+	public class NetworkPokeServer
+	{
+		private Socket _listener;
+		private readonly List<Socket> _clients = new List<Socket>();
+		ObservableCollection<string> _clientList;
+		private int _port;
+
+		private static string UpnpDescription = "Assembly Network Poking";
+
+		/// <summary>
+		/// Initializes a new instance of the <see cref="NetworkPokeServer"/> class.
+		/// A server will be created on the local machine on port 19002.
+		/// </summary>
+		public NetworkPokeServer(ObservableCollection<string> clientList)
+		{
+			_clientList = clientList;
+			// Discover UPnP support
+		}
+
+		public bool Listen(IPEndPoint endpoint)
+		{
+			try
+			{
+				_listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+
+				// Bind to our local endpoint
+				_listener.Bind(endpoint);
+				_port = endpoint.Port;
+				_listener.Listen(128); // Listen with a pending connection queue size of 128
+				NatUtility.DeviceFound += DeviceFound;
+				NatUtility.StartDiscovery();
+				return true;
+			}
+			catch (SocketException)
```

Catching this exception is going to hide the failure reason